### PR TITLE
.NET 7, bicep update, and partial parameters

### PR DESCRIPTION
**Pull request by nonik0**
## Description
Updates target to .net 7 and bicep dependency update with refactoring for use of DI for simplicity

Fixes an issue where a template fails to analyze when providing a parameters file with an incomplete set of parameters. Missing parameter values will now be generated.

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [x] The pull request does not introduce breaking changes.

### [General Guidelines](../CONTRIBUTING.md)
- [x] Title of the pull request is clear and informative.
- [x] Description of the pull request is clear and informative.
- [x] I have added myself to the 'assignees'.
- [x] I have added 'linked issues' if relevant.

### CODE REVIEW

File: src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs [hunk 1]
```diff
@@ -45,36 +66,17 @@ public class BicepTemplateProcessor
...
+                var serviceCollection = new ServiceCollection();
+                serviceCollection.AddBicepCore();
+
+                var services = serviceCollection.BuildServiceProvider();
+                BicepCompiler = services.GetRequiredService<BicepCompiler>();
             }
 
-            var compilation = new Compilation(featureProviderFactory, namespaceProvider, sourceFileGrouping, configurationManager, apiVersionProviderFactory, new LinterAnalyzer());
+            using var stringWriter = new StringWriter();
+            var compilation = BicepCompiler.CreateCompilation(new Uri(bicepPath)).Result; // TODO: sync call OK?
```
Comment (dotpaul) [inline src/Analyzer.BicepProcessor/BicepTemplateProcessor.cs:79]: >// TODO: sync call OK? [](http://example.com/codeflow?start=90&length=22)

I mean, if the calling method is synchronous, why not? :)

Is there an opportunity for Template Analyzer to do other work asynchronously while the Bicep compiler is compiling?

File: src/Analyzer.TemplateProcessor.UnitTests/ArmTemplateProcessorTests.cs [hunk 2]
```diff
@@ -1117,6 +1117,50 @@ public void ProcessTemplate_ValidTemplateUsingManagementGroupFunction_ProcessTem
...
+                        }
+                    }
+                ]
+            }";
+
+            var armTemplateProcessor = new ArmTemplateProcessor(templateJson);
+
+            JToken template = armTemplateProcessor.ProcessTemplate(parametersJson, null, generateMissingParameters: true);
+
+            Assert.AreEqual(2, template["parameters"].Count());
```
Comment (dotpaul) [inline src/Analyzer.TemplateProcessor.UnitTests/ArmTemplateProcessorTests.cs:1161]: >Assert.AreEqual(2, template["parameters"].Count()); [](http://example.com/codeflow?start=12&length=51)

Could also check the parameter values are as expected

File: src/Analyzer.TemplateProcessor.UnitTests/ArmTemplateProcessorTests.cs [hunk 3]
```diff
@@ -1117,6 +1118,61 @@ public void ProcessTemplate_ValidTemplateUsingManagementGroupFunction_ProcessTem
...
+                ]
+            }";
+
+            var armTemplateProcessor = new ArmTemplateProcessor(templateJson);
+
+            if (generateMissingParameters)
+            {
+                JToken template = armTemplateProcessor.ProcessTemplate(parametersJson, null, generateMissingParameters);
+                Assert.AreEqual(2, template["parameters"].Count());
+                Assert.IsNotNull(template["parameters"]["trafficRoutingMethod"]);
```
Comment (dotpaul) [inline src/Analyzer.TemplateProcessor.UnitTests/ArmTemplateProcessorTests.cs:1166]: Also check that trafficRoutingMethod is the "Priority" (the value specified)

File: src/Analyzer.TemplateProcessor.UnitTests/ArmTemplateProcessorTests.cs (hunk 2 above)
Comment (nonik0) [reply src/Analyzer.TemplateProcessor.UnitTests/ArmTemplateProcessorTests.cs]: done, also added negative case

**Review by dotpaul — APPROVED** [review]
:shipit:

File: src/Analyzer.Core/TemplateAnalyzer.cs [hunk 4]
```diff
@@ -134,7 +134,7 @@ private IEnumerable<IEvaluation> AnalyzeAllIncludedTemplates(string populatedTem
 
             try
             {
-                templatejObject = armTemplateProcessor.ProcessTemplate(parameters);
+                templatejObject = armTemplateProcessor.ProcessTemplate(parameters, parentContext.IsMainTemplate);
```
Comment (JohnathonMohr) [inline src/Analyzer.Core/TemplateAnalyzer.cs:137]: use named argument here for clarity

**Review by JohnathonMohr — APPROVED** [review]
In the past, getting the PowerShell environment to package correctly in the generated nupkg was complicated. There might be hidden breaking changes going from net6 to net7. Make sure to do lots of testing with PSRule execution specifically on the NuGet package/

**Pull request merged**